File: algo/Heap/KthLargest.cpp
class Heap {
    private:
    int * heap_arr = NULL;
    int heap_size;
    int heap_max_size;
    void swap(int index1, int index2)
    {
        int temp = heap_arr[index1];
        heap_arr[index1] = heap_arr[index2];
        heap_arr[index2] = temp; 
    }
    public:
    Heap(int heap_max)
    {
        heap_arr = new int[heap_max];
        heap_size = 0;
        heap_max_size = heap_max;
    }
    int Size()
    {
        return heap_size;
    }
    int Top()
    {
        return heap_arr[0];
    }
    void MinHeapify(int index)
    {
        int min_index = index;
        if(2*index+1 < heap_size && heap_arr[2*index+1] < heap_arr[min_index])
        {
            min_index = 2*index+1;
        }
        if(2*index+2 < heap_size && heap_arr[2*index+2] < heap_arr[min_index])
        {
            min_index = 2*index+2;
        }
        if(min_index != index)
        {
            swap(index, min_index);
            MinHeapify(min_index);
        }
        return;
    }
    int ExtractMin()
    {
        //cout << "heap size: " << heap_size << heap_arr[0] << heap_arr[1] << heap_arr[2] <<  endl;
        int minval = heap_arr[0];
        heap_size--;
        heap_arr[0] = heap_arr[heap_size];
        MinHeapify(0);
        return minval;
    }
    void BalancedInsert(int val)
    {
        heap_arr[0] = val;
        MinHeapify(0);
        return;
    }
    void Insert(int val)
    {
        int curr_index = heap_size;
        heap_arr[curr_index] = val;
        heap_size++;
        while(curr_index != 0)
        {
            int parent_index = (curr_index - 1)/2;
            if(heap_arr[curr_index] >= heap_arr[parent_index])
            {
                break;
            }
            else
            {
                swap(curr_index, parent_index);
            }
            curr_index = parent_index;
        }
        
        return;
    }
    
    
};
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        Heap *curr_heap = new Heap(k);
        //curr_heap->Insert(2);
        //curr_heap->Insert(1);
        //curr_heap->Insert(3);
        //cout << curr_heap->ExtractMin() << endl;
        //cout << curr_heap->ExtractMin() << endl;
        //cout << curr_heap->ExtractMin() << endl;
        for(int i = 0;i < nums.size();i++)
        {
            if(curr_heap->Size() < k)
            {
                curr_heap->Insert(nums[i]);
            }
            else
            {
                if(nums[i] > curr_heap->Top())
                {
                    //curr_heap->ExtractMin();
                    curr_heap->BalancedInsert(nums[i]);
                }
            }
        }
        return curr_heap->Top();
    }
};

File: algo/Heap/KthLargest.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
class Heap
	def initialize(heap_max)
		@heap_arr = Array.new(heap_max)
		@heap_size = 0
		@heap_max_size = heap_max
	end
	def swap(index1, index2)
		temp = @heap_arr[index1]
		@heap_arr[index1] = @heap_arr[index2]
		@heap_arr[index2] = temp
	end
	def Size()
		return @heap_size
	end
	def Top()
		return @heap_arr[0]
	end
	def MinHeapify(index)
		min_index = index
		if 2*index+1 < @heap_size and @heap_arr[2*index+1] < @heap_arr[min_index]
			min_index = 2*index+1
		end
		if 2*index+2 < @heap_size and @heap_arr[2*index+2] < @heap_arr[min_index]
			min_index = 2*index+2
		end
		if min_index != index
			swap(min_index, index)
			MinHeapify(min_index)
		end
	end
	def BalancedInsert(val)
		@heap_arr[0] = val
		MinHeapify(0)
	end
	def Insert(val)
        # puts(@heap_size)
		curr_index = @heap_size
		@heap_arr[curr_index] = val
		@heap_size += 1
		while curr_index != 0
			parent_index = (curr_index - 1)/2
            # print(parent_index, curr_index)
			if @heap_arr[parent_index] < @heap_arr[curr_index]
				break
			else
				swap(curr_index, parent_index)
			end
			curr_index = parent_index
		end
        # puts(@heap_size)
	end
end
def find_kth_largest(nums, k)
    # print(nums.size())
	curr_heap = Heap.new(k)
    for i in 0...nums.size()
        # puts(i)
    	if curr_heap.Size() < k
    		curr_heap.Insert(nums[i])
    	else
    		if nums[i] > curr_heap.Top()
    			curr_heap.BalancedInsert(nums[i])
            end
    	end
    end
    return curr_heap.Top()
end